auth/cache_test.go

```diff
@@ -0,0 +1,121 @@
+package auth
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/couchbase/sync_gateway/base"
+)
+
+func TestCache(t *testing.T) {
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyAuth)()
+	cache := NewCache(2)
+
+	k1 := key()
+	cache.Put(k1)
+	assert.Equal(t, 1, cache.Len())
+	assert.True(t, cache.Contains(k1))
+
+	k2 := key()
+	cache.Put(k2)
+	assert.Equal(t, 2, cache.Len())
+	assert.True(t, cache.Contains(k1) && cache.Contains(k2))
+
+	k3 := key()
+	cache.Put(k3)
+	assert.Equal(t, 2, cache.Len())
+	assert.True(t, cache.Contains(k1) && cache.Contains(k3) || cache.Contains(k2) && cache.Contains(k3))
+
+	k4 := key()
+	cache.Put(k4)
+	assert.Equal(t, 2, cache.Len())
+	assert.True(t, cache.Contains(k1) && cache.Contains(k4) || cache.Contains(k2) && cache.Contains(k4) || cache.Contains(k3) && cache.Contains(k4))
+}
+
+func key() (key string) {
+	uniq := time.Now().UnixNano()
+	key = fmt.Sprintf("k%d", uniq)
+	return key
+}
+
+func TestCacheRace(t *testing.T) {
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyAuth)()
+	var wg sync.WaitGroup
+	cache := NewCache(5)
+
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func(cache *Cache, i int) {
+			defer wg.Done()
+			key := fmt.Sprintf("k%d", i)
+			cache.Put(key)
+			base.Infof(base.KeyAuth, "Goroutine%v Put key: %v", i, key)
+		}(cache, i)
+
+		wg.Add(1)
+		go func(cache *Cache, i int) {
+			defer wg.Done()
+			key := fmt.Sprintf("k%d", i)
+			ok := cache.Contains(key)
+			base.Infof(base.KeyAuth, "Goroutine%v Contains: %v, key: %v", i, ok, key)
+		}(cache, i)
+	}
+	wg.Wait()
+	assert.Equal(t, 5, cache.Len())
+
+	cache.lock.RLock()
+	for key, value := range cache.cache {
+		base.Infof(base.KeyAuth, "Item: key: %v, value %v", key, value)
+	}
+	cache.lock.RUnlock()
+}
+
+func BenchmarkCachePut(b *testing.B) {
```

Comment:
It would be nice to have a benchmark for various scenarios of `compareHashAndPassword` so we can directly compare this implementation against the old one.

Of course there's some additional overhead to be able to do random-replacement, but we have no way of measuring that.

- unique correct hash/passwords (to test filling/replacing the cache)
- same correct hash/password (to test the fast-path of Contains)
- same incorrect hash/password (to verify we're still hitting the bcrypt slow-path and failing to insert)

Write this benchmark so it can be run against master, and then run the same benchmark on this branch. You can compare benchmark results by running the output of both through [benchstat](https://godoc.org/golang.org/x/perf/cmd/benchstat):

```shell
$ go test -run=- -bench=^BenchmarkFoo -benchmem -count=5 ./auth > old.txt
# ... change branch
$ go test -run=- -bench=^BenchmarkFoo -benchmem -count=5 ./auth > new.txt
$ benchstat old.txt new.txt
```

Reply:
Sure, that's a fair point. Added benchmark tests in the latest commit.